### DISCUSSION TITLE
[API Proposal]: try-catch with fallback

### DISCUSSION THREAD

**Discussion by PieterjanDeClippel** [dotnet/csharplang#8687] [Language Ideas]
### Background and motivation

Provide a shorter notation for commands with fail-cases

Prevents developers from writing the following

```csharp
namespace System.Collections.Generic;

void Example() {
    try {
        Method1();
    } catch(Exception ex1) {
        try {
            Method2();
        } catch(Exception ex2) {
            try {
                Method3();
            } catch(Exception ex3) {
                Console.WriteLine("Couldn't do it");
            }
        }
    }
}
```

### API Proposal

```csharp
namespace System.Collections.Generic;

void Example() {
    try {
        Method1();
    } fallback(Exception ex1) {
        Method2();
    } fallback(Exception ex2) {
        Method3();
    } catch(Exception ex3) {
        Console.WriteLine("Couldn't do it");
    }
}
```

### API Usage

```csharp
namespace System.Collections.Generic;

void Example() {
    try {
        Method1();
    } fallback(Exception ex1) {
        Method2();
    } fallback(Exception ex2) {
        Method3();
    } catch(Exception ex3) {
        Console.WriteLine("Couldn't do it");
    }
}
```

### Risks

Variables from a former scope are not available in subsequent `fallback` blocks

**Comment by huoyaoyuan** [comment]
This is a C# language proposal, not runtime API proposal. It should go to https://github.com/dotnet/csharplang/discussions .

This is very unlikely to be considered because it's not a common or encouraged pattern. Related to https://github.com/dotnet/csharplang/discussions/1015 and https://github.com/dotnet/csharplang/discussions/4676 .

**Comment by Zintom** [comment]
I feel like this could be implemented at an application level rather than updating the runtime/language.

There are libraries that already do this.
